rmf_robot_sim_common/src/slotcar_common.cpp

```diff
@@ -653,54 +768,108 @@ SlotcarCommon::UpdateResult SlotcarCommon::update_ackermann(
-        result.speed = 0;
-        /*
-        // Apply speed limit if one is present in the trajectory
-        if (traj.approach_speed_limit.has_value())
+        // When near the last waypoint, directly head towards it.
+        target = trajectory.at(_traj_wp_idx).pose.translation().head<2>();
+      }
+      else
+      {
+        // Otherwise, head towards the closest point on the path which
+        // is at least one lookahead distance away from the vehicle.
+
+        // points A and B represent the start and end of the current path segment.
+        Eigen::Vector2d point_A;
+        if (_traj_wp_idx == 0)
         {
-          result.speed = traj.approach_speed_limit.value();
-          result.max_speed = traj.approach_speed_limit.value();
+          point_A = _pose.translation().head<2>();
+        }
+        else
+        {
+          point_A = trajectory.at(_traj_wp_idx-1).pose.translation().head<2>();
+        }
+        auto point_B = trajectory.at(_traj_wp_idx).pose.translation().head<2>();
+
+        // Find the intersection of the circle around the vehicle
+        // with the path segment.
+        auto intersections = line_circle_intersections(
+          point_A,
+          point_B,
+          _pose.translation()(0),
+          _pose.translation()(1),
+          _lookahead_distance
+        );
+
+        if (intersections.size() == 0)
+        {
+          // No intersections; head towards closest point on the path segment.
+          auto point_P = _pose.translation().head<2>();
+          target = get_closest_point_on_line_segment(point_A, point_B, point_P);
+        }
+        else if (intersections.size() == 1)
+        {
+          target = intersections.at(0);
+        }
+        else if (intersections.size() == 2)
+        {
+          // Select the intersection closer to B.
+          auto distance_0 = (point_B.head<2>() - intersections.at(0)).norm();
+          auto distance_1 = (point_B.head<2>() - intersections.at(1)).norm();
+          target = intersections.at(distance_0 < distance_1 ? 0 : 1);
         }
-        */
       }
-      else
+
+      auto target_3d = Eigen::Vector3d(target(0), target(1), 0);
+      _pursuit_state.lookahead_point(0) = target_3d(0);
+      _pursuit_state.lookahead_point(1) = target_3d(1);
+      _pursuit_state.lookahead_point(2) = target_3d(2);
+      _pursuit_state.traj_wp_idx = _traj_wp_idx;
+
+      Eigen::Vector3d d_target = target_3d - _pose.translation();
+
+      result.v = d_target.norm();
+
+      auto goal_heading = compute_heading(trajectory.at(_traj_wp_idx).pose);
+      double dir = 1.0;
+      result.w = compute_change_in_rotation(
+        current_heading, d_target, &goal_heading, &dir);
+
+      // As turning yaw increases, slow down more, to a minimum of 20%.
+      double turning = fabs(result.w) / M_PI;
+      double slowdown = std::min(0.2, 2 * turning);
+      result.target_linear_speed_now = result.max_speed.value() *
+        (1 - slowdown);
```

Comment:
My mistake, should have been 0.8 instead of 0.2. Updated it to 
`double slowdown = std::min(0.8, turning);` so that:

- Set slowdown to be the minimum of turning and 0.8.
- Scale speed to `1-slowdown` 
- In most cases, speed is slowed proportionally to turning,
- except when turning > 0.8 (144 degrees), maintain a speed of at least 20% so as to not stop the vehicle.